xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -745,6 +1002,138 @@ void onRejected(String rejectedVersion, long rejectedTime, String rejectedDetail
     private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
       watcher.onChanged(update);
     }
+
+    void setCpc(ControlPlaneClient controlPlaneClient) {
+      checkNotNull(controlPlaneClient, "controlPlaneClient");
+      this.controlPlaneClient = controlPlaneClient;
+    }
+  }
+
+  private class ResponseHandler implements XdsResponseHandler {
+    final ServerInfo serverInfo;
+
+    ResponseHandler(ServerInfo serverInfo) {
+      this.serverInfo = serverInfo;
+    }
+
+    @Override
+    public void handleResourceResponse(
+        XdsResourceType<?> xdsResourceType, ServerInfo serverInfo, String versionInfo,
+        List<Any> resources, String nonce, ProcessingTracker processingTracker) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean inRetry,
+                                   ControlPlaneClient cpcThatClosed) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      ControlPlaneClient cpcForThisStream = serverCpClientMap.get(serverInfo);
```

Comment:
Isn't cpcThatClosed the same instance as cpcForThisStream?

Reply:
Good point, consolidated them